From ce508ba4620aacac81fb5a1db70c2d4e03125d69 Mon Sep 17 00:00:00 2001
From: Eberhard Beilharz <eb1@sil.org>
Date: Mon, 9 Jan 2012 12:32:58 +0100
Subject: [PATCH] bug-xamarin-2787: release capture when clicking LinkLabel

When clicking on a LinkLabel we need to release the mouse capture
otherwise the UI might lock up.

This fixes Xamarin bug 2787.

Change-Id: Ibcc0fe8239768e3945a3592c789fd2730d877390
---
 .../System.Windows.Forms/LinkLabel.cs              |   22 ++++++++++----------
 1 files changed, 11 insertions(+), 11 deletions(-)

Index: fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/LinkLabel.cs
===================================================================
--- fieldworks-mono.orig/mcs/class/Managed.Windows.Forms/System.Windows.Forms/LinkLabel.cs
+++ fieldworks-mono/mcs/class/Managed.Windows.Forms/System.Windows.Forms/LinkLabel.cs
@@ -464,22 +464,22 @@
 		{
 			if (!Enabled) return;
 
-			base.OnMouseUp (e);
+			if (active_link != null && this.Capture && ((e.Button & MouseButtons.Left) != 0)) {
+				this.Capture = false;
+				Link clicked_link = (PointInLink (e.X, e.Y) == active_link) ? active_link : null;
 
-			if (active_link == null)
-				return;
+				active_link.Active = false;
+				active_link = null;
 
-			Link clicked_link = (PointInLink (e.X, e.Y) == active_link) ? active_link : null;
-
-			active_link.Active = false;
-			active_link = null;
-
-			if (clicked_link != null)
+				if (clicked_link != null)
 #if NET_2_0
-				OnLinkClicked (new LinkLabelLinkClickedEventArgs (clicked_link, e.Button));
+					OnLinkClicked (new LinkLabelLinkClickedEventArgs (clicked_link, e.Button));
 #else
-				OnLinkClicked (new LinkLabelLinkClickedEventArgs (clicked_link));
+					OnLinkClicked (new LinkLabelLinkClickedEventArgs (clicked_link));
 #endif
+			}
+
+			base.OnMouseUp(e);
 		}
 
 		protected override void OnPaint (PaintEventArgs e)
